x-pack/legacy/plugins/alerting/server/builtin_alert_types/index_threshold/index.ts

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { i18n } from '@kbn/i18n';
+import { Logger } from '../../../../../../../src/core/server';
+
+import { AlertType, AlertExecutorOptions } from '../../types';
+import { Params, ParamsSchema } from './params';
+import { BaseActionContext, addMessages } from './action_context';
+
+import { runQuery } from '../../../../../../plugins/alerting_index_threshold/server';
+
+export const ID = '.index-threshold';
+
+interface GetAlertTypeParams {
+  logger: Logger;
+}
+
+const ActionGroup = 'threshold met';
+const CompareFns = getCompareFns();
+
+export function getAlertType(getAlertTypeParams: GetAlertTypeParams): AlertType {
+  const { logger } = getAlertTypeParams;
+
+  const alertTypeName = i18n.translate('xpack.alerting.builtin.indexThresholdTitle', {
+    defaultMessage: 'Index Threshold',
+  });
+
+  return {
+    id: ID,
+    name: alertTypeName,
+    actionGroups: [ActionGroup],
+    validate: {
+      params: ParamsSchema,
+    },
+    executor,
+  };
+
+  async function executor(options: AlertExecutorOptions) {
+    const { alertId, name, services } = options;
+    const params: Params = options.params as Params;
+
+    const compareFn = CompareFns[params.comparator];
+    if (compareFn == null) {
+      throw new Error(`invalid comparator specified: "${params.comparator}"`);
+    }
+
+    const callCluster = services.callCluster;
+    const date = new Date().toISOString();
+    const queryParams = {
+      index: params.index,
+      timeField: params.timeField,
+      aggType: params.aggType,
+      aggField: params.aggField,
+      groupField: params.groupField,
+      groupLimit: params.groupLimit,
+      date,
+      interval: params.window,
+      window: params.window,
+      intervalsBefore: 1,
+      intervalsAfter: 0,
+    };
+    const result = await runQuery({ logger, callCluster, queryParams });
```

Comment:
It looks like the `runQuery` function is the only thing we need to replace in order to reuse all of this logic for threshold alerting on other things (e.g. metrics). Maybe we can produce a factory that takes in a `(params: Object) => QueryResult` function to easily register threshold alert types with common functionality?

Reply:
Ya, that would be an interesting way to try to make this reusable.  The `runQuery` function is also used to generate data for the UI preview, and perhaps history view.  So if we spec it just right, it could be possible to supply a `runQuery` function that would get used in all those environments.  I think the params and probably need to be configurable, but the result (at least right now) for the query is simply an array of `[dateString, number]` tuples.

I'd like to get this merged into master before looking at generalizing it more, but I'll keep this in mind while finishing this.

I should mention that 1/4 of the content of this code is setting the context for the `scheduleActions()` call - and most of the things being set are generic off the alert, params, etc.  We've talked about providing these for "free" to action groups, so that we'd still have the `context` we currently support, for alertType-specific things, but you would also somehow have access to everything in the alert as well, based on some other variable.